File: app/src/main/java/org/worshipsongs/service/PresentationScreenService.java
package org.worshipsongs.service;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.DialogInterface;
import android.media.MediaRouter;
import android.os.Build;
import android.util.Log;
import android.view.Display;
import android.view.View;

import org.worshipsongs.dialog.RemoteSongPresentation;
import org.worshipsongs.domain.Setting;
import org.worshipsongs.domain.Song;
import org.worshipsongs.utils.CommonUtils;

/**
 * Author : Madasamy
 * Version : 3.x
 */

public class PresentationScreenService
{
    private UserPreferenceSettingService preferenceSettingService = new UserPreferenceSettingService();
    private Context context;
    private RemoteSongPresentation remoteSongPresentation;
    private PresentationScreenService.DefaultMediaRouterCallBack songMediaRouterCallBack = new PresentationScreenService.DefaultMediaRouterCallBack();
    private MediaRouter mediaRouter;

    public PresentationScreenService()
    {

    }

    public PresentationScreenService(Context context)
    {
        this.context = context;
        mediaRouter = (MediaRouter) context.getSystemService(Context.MEDIA_ROUTER_SERVICE);
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    public void onResume()
    {
        if (isJellyBean() && CommonUtils.isProductionMode()) {
            mediaRouter.addCallback(MediaRouter.ROUTE_TYPE_LIVE_VIDEO, songMediaRouterCallBack);
            updatePresentation();
        }
    }


    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    public void onPause()
    {
        if (isJellyBean() && CommonUtils.isProductionMode()) {
            mediaRouter.removeCallback(songMediaRouterCallBack);
        }
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    public void onStop()
    {
        try {
            if (remoteSongPresentation != null) {
                remoteSongPresentation.cancel();
                remoteSongPresentation = null;
            }
        } catch (Exception ex) {
            Log.e(PresentationScreenService.class.getSimpleName(), "Error occurred while dismiss remote display");
        }
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    private class DefaultMediaRouterCallBack extends MediaRouter.SimpleCallback
    {

        @Override
        public void onRouteSelected(MediaRouter router, int type, MediaRouter.RouteInfo info)
        {
            updatePresentation();

        }

        @Override
        public void onRouteUnselected(MediaRouter router, int type, MediaRouter.RouteInfo info)
        {
            updatePresentation();

        }

        @Override
        public void onRoutePresentationDisplayChanged(MediaRouter router, MediaRouter.RouteInfo info)
        {
            updatePresentation();
        }

    }

    private final DialogInterface.OnDismissListener remoteDisplayDismissListener =
            new DialogInterface.OnDismissListener()
            {
                @Override
                public void onDismiss(DialogInterface dialog)
                {
                    if (dialog == remoteSongPresentation) {
                        remoteSongPresentation = null;
                    }

                }
            };

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    private void updatePresentation()
    {
        try {
            Display selectedDisplay = getSelectedDisplay();
            if (remoteSongPresentation != null && remoteSongPresentation.getDisplay() != selectedDisplay) {
                if (remoteSongPresentation.isShowing()) {
                    remoteSongPresentation.dismiss();
                }
                remoteSongPresentation = null;
            }

            if (remoteSongPresentation == null && selectedDisplay != null) {
                remoteSongPresentation = new RemoteSongPresentation(context, selectedDisplay);
                remoteSongPresentation.setOnDismissListener(remoteDisplayDismissListener);
                remoteSongPresentation.show();
            }
            if (remoteSongPresentation != null && selectedDisplay != null) {
                if (Setting.getInstance().getSong() != null) {
                    showNextVerse(Setting.getInstance().getSong(), Setting.getInstance().getSlidePosition());
                }
            }
        } catch (Exception ex) {
            remoteSongPresentation = null;
            Log.e(PresentationScreenService.class.getSimpleName(), "Error occurred while presenting remote display" + ex);
        }
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    private Display getSelectedDisplay()
    {
        MediaRouter.RouteInfo selectedRoute = mediaRouter.getSelectedRoute(
                MediaRouter.ROUTE_TYPE_LIVE_VIDEO);
        Display selectedDisplay = null;
        if (selectedRoute != null) {
            selectedDisplay = selectedRoute.getPresentationDisplay();
        }
        return selectedDisplay;
    }

    private boolean isJellyBean()
    {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1;
    }

    public void showNextVerse(Song song, int position)
    {
        try {
            if (remoteSongPresentation != null) {
                remoteSongPresentation.setVerseVisibility(View.VISIBLE);
                remoteSongPresentation.setImageViewVisibility(View.GONE);
                remoteSongPresentation.setVerse(song.getContents().get(position));
                remoteSongPresentation.setSongTitleAndChord(song.getTitle(), song.getChord(), preferenceSettingService.getPresentationPrimaryColor());
                remoteSongPresentation.setAuthorName(song.getAuthorName(), preferenceSettingService.getPresentationPrimaryColor());
                remoteSongPresentation.setSlidePosition(position, song.getContents().size(), preferenceSettingService.getPresentationPrimaryColor());
                Setting.getInstance().setSong(song);
                Setting.getInstance().setSlidePosition(position);
            }
        } catch (Exception e) {
            Log.e(PresentationScreenService.class.getSimpleName(), "Error occurred while presenting song content" + e);
        }
    }

    public RemoteSongPresentation getPresentation()
    {
        return remoteSongPresentation;
    }

}

File: app/src/main/java/org/worshipsongs/service/PresentationScreenService.kt
package org.worshipsongs.service

import android.annotation.TargetApi
import android.content.Context
import android.content.DialogInterface
import android.os.Build
import android.util.Log
import android.view.View
import androidx.mediarouter.media.MediaControlIntent
import androidx.mediarouter.media.MediaRouteSelector
import androidx.mediarouter.media.MediaRouter
import org.worshipsongs.dialog.RemoteSongPresentation
import org.worshipsongs.domain.Setting
import org.worshipsongs.domain.Song
import org.worshipsongs.utils.CommonUtils

/**
 * @author: Madasamy
 * @version: 3.x
 */

class PresentationScreenService
{
    private val preferenceSettingService = UserPreferenceSettingService()
    private var context: Context? = null
    var presentation: RemoteSongPresentation? = null
    private var songMediaRouterCallBack = DefaultMediaRouterCallBack()
    private var mediaRouter: MediaRouter? = null

    private val remoteDisplayDismissListener = DialogInterface.OnDismissListener { dialog ->
        if (dialog == presentation)
        {
            presentation = null
        }
    }

    private val isJellyBean: Boolean get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    constructor(context: Context)
    {
        this.context = context
        mediaRouter = MediaRouter.getInstance(context)
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    private inner class DefaultMediaRouterCallBack : MediaRouter.Callback()
    {

        override fun onRouteSelected(router: MediaRouter, info: MediaRouter.RouteInfo)
        {
            updatePresentation()

        }

        override fun onRouteUnselected(router: MediaRouter, info: MediaRouter.RouteInfo)
        {
            updatePresentation()

        }

        override fun onRoutePresentationDisplayChanged(router: MediaRouter, info: MediaRouter.RouteInfo)
        {
            updatePresentation()
        }

    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    private fun updatePresentation()
    {
        try
        {
            val selectedRoute = mediaRouter!!.getSelectedRoute()
            val selectedDisplay = selectedRoute.presentationDisplay
            if (presentation != null && presentation!!.display != selectedDisplay)
            {
                if (presentation!!.isShowing)
                {
                    presentation!!.dismiss()
                }
                presentation = null
            }
            Log.i(this.javaClass.simpleName, "Presentation: $presentation")
            if (presentation == null && selectedDisplay != null)
            {
                presentation = RemoteSongPresentation(context!!, selectedDisplay)
                presentation!!.setOnDismissListener(remoteDisplayDismissListener)
                presentation!!.show()
            }
            if (presentation != null && selectedDisplay != null)
            {
                if (Setting.instance.song != null)
                {
                    showNextVerse(Setting.instance.song, Setting.instance.slidePosition)
                }
            }
        } catch (ex: Exception)
        {
            presentation = null
            Log.e(PresentationScreenService::class.java.simpleName, "Error occurred while presenting remote display", ex)
        }

    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    fun onResume()
    {
        if (isJellyBean && CommonUtils.isProductionMode)
        {
            mediaRouter!!.addCallback(MediaRouteSelector.Builder().addControlCategory(MediaControlIntent.CATEGORY_LIVE_VIDEO).build(), songMediaRouterCallBack)
            updatePresentation()
        }
    }


    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    fun onPause()
    {
        if (isJellyBean && CommonUtils.isProductionMode)
        {
            mediaRouter!!.removeCallback(songMediaRouterCallBack!!)
        }
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    fun onStop()
    {
        try
        {
            if (presentation != null)
            {
                presentation!!.cancel()
                presentation = null
            }
        } catch (ex: Exception)
        {
            Log.e(PresentationScreenService::class.java.simpleName, "Error occurred while dismiss remote display", ex)
        }

    }

    fun showNextVerse(song: Song?, position: Int)
    {
        try
        {
            if (presentation != null)
            {
                presentation!!.setVerseVisibility(View.VISIBLE)
                presentation!!.setImageViewVisibility(View.GONE)
                presentation!!.setVerse(song!!.contents!![position])
                presentation!!.setSongTitleAndChord(song.title!!, song.chord!!, preferenceSettingService.presentationPrimaryColor)
                presentation!!.setAuthorName(song.authorName!!, preferenceSettingService.presentationPrimaryColor)
                presentation!!.setSlidePosition(position, song.contents!!.size, preferenceSettingService.presentationPrimaryColor)
                Setting.instance.song = song
                Setting.instance.slidePosition = position
            }
        } catch (e: Exception)
        {
            Log.e(PresentationScreenService::class.java.simpleName, "Error occurred while presenting song content$e")
        }
    }

}
